### ISSUE TITLE
Implement nemo ofx variables for cmip6 cmorization

### ISSUE THREAD

**Issue by treerink** [EC-Earth/ece2cmor3#249]
This issue has been mentioned in https://github.com/EC-Earth/ece2cmor3/issues/202#issuecomment-404200479 as sub issue.

**Comment by goord** [comment]
Example variables: thkcello, masscello etc. in CMIP6_Ofx.json

The problem is these variables are not in our output netcdf files, but perhaps in the ocean mesh definition files of EC-Earth?

**Comment by treerink** [comment]
I am afraid we currently never load the Ofx variables in the taskloader:
`````
        if sheetname.lower() in ["notes", "fx", "Ofx"]:
            continue
`````

**Comment by goord** [comment]
Hi @treerink I committed support code for the Ofx-variables in the Ofx-support branch. I also removed the exception above for the Ofx sheetname in this branch. Can you test your full cmorization once more, and maybe also some of your drq-to-nemo-filedef scripts?

**Comment by treerink** [comment]
In cbb47bb I activated also table "fx" in the Ofx-branch, because the tests require the full chain: genecec + run EC-Earth3 + cmorization, so I prefer to do it in one go. I saw you added the "fx" case in `nemo2cmor.py` already. I can see potentially seven NEMO "Ofx" variables will be added, for "fx" I doubt if any NEMO variable is added (but "fx" is bad grep string to check). 

@goord nothing against including "fx" in the test loop on forehand, right?

**Comment by treerink** [comment]
Enabling the "fx" table appeared to be not possible yet, because of the IFS part.. No ppt files can be produced with this because `drq2ppt.py` crashes, see #335.

**Comment by treerink** [comment]
I tried "Ofx" variable `deptho`, but get the following error:
`````
INFO:ece2cmor3.nemo2cmor:Start cmorization of deptho in table Ofx
INFO:ece2cmor3.nemo2cmor:Creating time axes for table Ofx from data in cmip6-ec-earth-output/t214/001/t214_1y_19900101_19901231_opa_grid_T_3D.nc...
INFO:ece2cmor3.nemo2cmor:Creating depth axes for table Ofx from data in cmip6-ec-earth-output/t214/001/t214_1y_19900101_19901231_opa_grid_T_3D.nc ...
INFO:ece2cmor3.nemo2cmor:Creating depth axes for table Ofx using file cmip6-ec-earth-output/t214/001/t214_1y_19900101_19901231_opa_grid_T_3D.nc...
INFO:ece2cmor3.nemo2cmor:Creating extra axes for table Ofx from data in cmip6-ec-earth-output/t214/001/t214_1y_19900101_19901231_opa_grid_T_3D.nc ...
INFO:ece2cmor3.nemo2cmor:Creating extra axes for table Ofx using file cmip6-ec-earth-output/t214/001/t214_1y_19900101_19901231_opa_grid_T_3D.nc...

C Traceback:
! In function: cmor_variable
! 

!!!!!!!!!!!!!!!!!!!!!!!!!
!
! Error: You defined variable 'deptho' (table Ofx) with axis id 'depth_coord' which is not part of this variable according to your table, it says: ( latitude longitude )
!
!!!!!!!!!!!!!!!!!!!!!!!!!

Traceback (most recent call last):
  File "/usr/people/reerink/anaconda2/envs/ece2cmor3/bin/ece2cmor", line 11, in <module>
    load_entry_point('ece2cmor3', 'console_scripts', 'ece2cmor')()
  File "/usr/people/reerink/cmorize/ece2cmor3/ece2cmor3/ece2cmor.py", line 138, in main
    ece2cmorlib.perform_nemo_tasks(args.datadir, args.exp, refdate)
  File "/usr/people/reerink/cmorize/ece2cmor3/ece2cmor3/ece2cmorlib.py", line 193, in perform_nemo_tasks
    nemo2cmor.execute(nemo_tasks)
  File "/usr/people/reerink/cmorize/ece2cmor3/ece2cmor3/nemo2cmor.py", line 108, in execute
    execute_netcdf_task(dataset, task)
  File "/usr/people/reerink/cmorize/ece2cmor3/ece2cmor3/nemo2cmor.py", line 147, in execute_netcdf_task
    varid = create_cmor_variable(task, dataset, axes)
  File "/usr/people/reerink/cmorize/ece2cmor3/ece2cmor3/nemo2cmor.py", line 209, in create_cmor_variable
    original_name=str(srcvar))
  File "/usr/people/reerink/anaconda2/envs/ece2cmor3/lib/python2.7/site-packages/cmor/pywrapper.py", line 546, in variable
    missing_value, tolerance, positive, original_name, history, comment)
_cmor.CMORError: Problem with 'cmor.variable'. Please check the logfile (if defined).
`````

**Comment by goord** [comment]
'deptho' in our output should not be associated with the variable on the Ofx table: the former gives the layer depth w.r.t. the surface, the latter is the ocean bathymetry.

**Comment by treerink** [comment]
Ok so this means for `deptho` we seem to have a wrong identification in the ping file (maybe due to a data request change). So we have to figure out the NEMO code name for bathymetry, or else if we can't find it we skip it. It probably could also be obtained from the initial files.

**Comment by goord** [comment]
I noticed that the requested deptho is defined in the ```bathy_meter.nc``` file in the nemo grid definition files in EC-Earth.

**Comment by treerink** [comment]
Only `depth0` needs still to be addressed indeed. If I take `depth0` off from the `varlist-nemo-all.json` and run:
`````
 ece2cmor cmip6-ec-earth-output/t230/001/ --exp t230 --oce --conf ece2cmor3/resources/metadata-templates/cmip6-CMIP-piControl-metadata-template.json --vars ece2cmor3/resources/test-data-request/varlist-nemo-all.json --odir cmor-test-all-230 >& log-cmip6-cmorizing-nemo-test-all-t230 &
`````
`ece2cmor` ends with zero errors and zero warnings, and 171 variables have been produced. That means the rest of the Ofx variables went fine. And for instance hfx has proper field values now.

**Comment by goord** [comment]
So to summarize, 'deptho' is bathymetry, and it can be found in the nemo input files, which is a bit cumbersome for ece2cmor3 because now it has to know where that data is. Another possibility would be to include the bathymetries in our own resources, but I'm not sure whether we're allowed to do so...

**Comment by klauswyser** [comment]
The bathymetry is even more tricky: the bathymetry.nc file that we provide to NEMO is not exactly the bathymetry that NEMO uses later, a few land gridpoints are set to water to make sure ocean currents can pass through narrow straits, and these points also receive a depth. It therefore makes more sense to use the depth from the mesh_mask file that is saved in NEMO, but we usually suppress the writing of this file. 

A not so nice but pragmatic approach would be to create the deptho file off-line once for each ocean resolution, and then manually add this file to the ofx directory.

**Comment by goord** [comment]
Thanks for pointing that out Klaus, I remember using the mesh_mask.nc file for the primavera ocean grid description. I will test tomorrow whether I can make ece2cmor3 read this file and cmorize deptho from it.

**Comment by treerink** [comment]
The `deptho`can be done in the post process step (with clim and dec ones), targeting for release 1.0 is enough for that. The rest is solved already.

**Comment by zklaus** [comment]
There maybe a couple of more issues with Ofx data. Please let me know if you prefer separate issues here:

- Due to our use of pressure coordinates, the cell thickness in m varies over time. As a consequence, the following variables should not be stored in Ofx, but instead in Omon and Odec: `masscello`, `thkcello`, `volcello`.
- The sea area fraction `sftof` should be in Ofx, but then it shouldn't contain information about ice since that is time dependent. However, looking at [the current output](https://github.com/EC-Earth/ece2cmor3/files/2791597/ncview.sftof.pdf) this seems to happen. Since we don't use fractional ocean cells in the model, it would probably be good to derive this field simply from the land-sea-mask, setting it to 1 over the sea and masked over land.

**Comment by goord** [comment]
Ah thanks for your input @zklaus, I didn't realize they were also in the monthly and decadal tables.

**Comment by zklaus** [comment]
I think it is even different: If your model uses z-levels they are fixed and should be in Ofx and not in (Omon, Odec). If they are time-dependent, for example because you use pressure or hybrid levels like we do, they should not be in Ofx, but only in (Omon, Odec). The documentation is a bit confusing here, but if you look at the data request you find:

[(Ofx, `thkcello`)](http://clipc-services.ceda.ac.uk/dreq/u/bab9bd00-e5dd-11e5-8482-ac72891c3257.html):
```
processing: If this field is time-dependent then save it instead as one of your Omon fields (see the Omon table)
```
[(Ofx, `masscello`)](http://clipc-services.ceda.ac.uk/dreq/u/baa3ea2a-e5dd-11e5-8482-ac72891c3257.html)
```
description: [...] More generally, masscello is time dependent and reported as part of Omon.
```
[(Ofx, `volcello`)](http://clipc-services.ceda.ac.uk/dreq/u/babcc39c-e5dd-11e5-8482-ac72891c3257.html)
```
description: [...] If this field is time-dependent then save it instead as one of your Omon and Odec fields
```
the operative word here being instead.
In the data request, all three appear in Ofx, Omon, Odec, but volcello additionally appears in Oyr.
So, yeah.

**Comment by goord** [comment]
So @treerink I suggest we move (Ofx, thkcello), (Ofx, masscello) and (Ofx, volcello) to the ignored variables list, they are time-dependent for nemo, and we take off the release 1 label. We will add the bathymetry extraction to a later minor release.

**Comment by goord** [comment]
cmorizing deptho from bathy_meter file is now supported in the deptho-bathy branch. @treerink I leave it to you to add the thkcello, masscello and volcello in Ofx to the ignored list

**Comment by treerink** [comment]
`volcello` is in ignored list already. I will move `deptho` to ignored list, to prevent further confusion and it unnecessary production, because it will be taken from the `bathy_meter.nc` file.

Checking `thkcello` and `masscello.`

**Comment by treerink** [comment]
The issue with `deptho` is that is actually identified by Shaconemo in the ping file, as ec-earth community we agreed that this is a miss identification and that it should come directly from the `bathy_meter.nc` file. I could make the change at the root, and change in our repository the `deptho` identification into "dummy" in the `ping_ocean_DR1.00.27.xml` file on top of the Shaconemo version and rerun all the steps, but at this stage I prefer to leave the Shaconemo identification as it is for this variable and just remove the `deptho` at the highest level in `genecec` so it won't end up in the final `file_def` files.

**Comment by treerink** [comment]
The `thkcello` and `masscello` are also asked at the Omon table. So I will  just add the Ofx combinations with `thkcello` and `masscello` to the ignored list, right?

**Comment by goord** [comment]
Yes

**Comment by treerink** [comment]
@goord I tried to cmorize just deptho but the `bathy_meter.nc` seems not available. I thought the idea is to upload this netcdf file in the `ece2cmor3` repository, right?

**Comment by goord** [comment]
Hi Thomas, ece2cmor3 looks 3 directories up from your input data directory, which should be your EC-Earth run directory (unless you are using symlinks). Do you see the file there?

If you have a different location, you can export the environment variable ``ECE2CMOR3_NEMO_BATHY_METER`` to the file path of the Bathy_meter.nc and this should be then picked up by ece2cmor3.

**Comment by treerink** [comment]
Hmm can't find it at all in the ec-earth svn repository. Is it created at some initial step? But ven then I can't find it with `find`. What should be the path of its directory?

**Comment by goord** [comment]
You should find it in the initial data directory of Philippe on ecfs. The problem with including it is the resolution, ece2cmor has been unaware of this

**Comment by treerink** [comment]
Ok copying the file for instance in my home and using in my bash shell:
`````
export ECE2CMOR3_NEMO_BATHY_METER=${HOME}/bathy_meter.nc
`````
does work. A best way of working for this specific is addressed in #424.

Closing this issue, as the Ofx variables are working correctly now.
